compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala

```diff
@@ -469,10 +469,12 @@ trait QuotesAndSplices {
     val matchModule = if tree.quoted.isTerm then defn.QuoteMatching_ExprMatch else defn.QuoteMatching_TypeMatch
     val unapplyFun = qctx.asInstance(defn.QuoteMatchingClass.typeRef).select(matchModule).select(nme.unapply)
 
-    UnApply(
-      fun = unapplyFun.appliedToTypeTrees(typeBindingsTuple :: TypeTree(patType) :: Nil),
-      implicits = quotedPattern :: Nil,
-      patterns = splicePat :: Nil,
-      proto = quoteClass.typeRef.appliedTo(replaceBindings(quoted1.tpe) & quotedPt))
+    Typed(
+      UnApply(
+        fun = unapplyFun.appliedToTypeTrees(typeBindingsTuple :: TypeTree(patType) :: Nil),
+        implicits = quotedPattern :: Nil,
+        patterns = splicePat :: Nil),
+      TypeTree(quoteClass.typeRef.appliedTo(replaceBindings(quoted1.tpe) & quotedPt))
+    ).annotated(New(ref(defn.UncheckedAnnot.typeRef)))
```

Comment:
What's this?  First of all it's apparently meant to be `New(defn.UncheckedAnnot.typeRef, Nil)` (there's an overarching ticket on this too).  But also, why is this thrown in here?

Reply:
Some of the quote patterns did not compile due to an unchecked warning/error in the library. I'm not sure yet if this is the proper fix. I added it to see if I uncover more issues not related to quote pattern matching.

Reply:
The reason why this might be ok is that the pattern complains that it cannot check at runtime an `Expr[T]` because `T` is erased, but the quoted pattern matches on the trees that do have the unerased info.

Reply:
After looking into it, we do need this extra type ascription to support the following code
```scala
   case x @ '{ ...: T } => x: Expr[T]
```
This code gets compiled to
```scala
  case quotes.ExprMatch[Tup1, Tup2](...)(using '{ ...: T }) => 
```

where `Tup1` and `Tup2` help type the extracted variables.

The unapply is defined as
```scala
def unapply[TypeBindings <: Tuple, Tup <: Tuple](scrutinee: Expr[Any])(using pattern: Expr[Any]): Option[Tup]
```

The issue is that this takes as a scrutinee an `Expr[Any]` but we know that it will only match if it is an `Expr[T]`. To convey this extra knowledge we add the unchecked type ascription around the `unapply`.

We could do it without an unchecked ascription using a `TypeTest` instead of the ascription. This would be less performant.

Reply:
Thanks for the explanation, could you add a comment in the code about this?

Reply:
I don't fully follow the logic, but thanks for digging into this some more.  I trust your judgement, and we can always revisit this later.

I wonder whether this change allows for these special cases to be dropped:

https://github.com/lampepfl/dotty/blob/dc60c4010624edb29193f8535ae4988c454e27a8/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala#L873-L881

Reply:
Yes, we can drop this special case. Thank you.

Reply:
It seems I was mistaken, we seem to need it. I will have a deeper look at this later.

Reply:
> Thanks for the explanation, could you add a comment in the code about this?

I updated the docstring to contain the new details